libxbot-service/src/Scheduler.cpp

```diff
@@ -0,0 +1,77 @@
+#include <xbot-service/Lock.hpp>
+#include <xbot-service/Scheduler.hpp>
+#include <xbot-service/portable/system.hpp>
+
+xbot::service::Scheduler::~Scheduler() {
+  thread::deinitialize(thread_);
+  mutex::deinitialize(&state_mutex_);
+}
+
+bool xbot::service::Scheduler::Start() {
+  stopped_ = false;
+
+  if (!mutex::initialize(&state_mutex_)) {
+    return false;
+  }
+
+#ifdef XBOT_ENABLE_STATIC_STACK
+  if (!thread::initialize(thread_, ThreadLoopHelper, this, thread_stack_, sizeof(thread_stack_), "Scheduler"))
+#else
+  if (!thread::initialize(thread_, ThreadLoopHelper, this, nullptr, 0, "Scheduler"))
+#endif  // XBOT_ENABLE_STATIC_STACK
+  {
+    return false;
+  }
+
+  return true;
+}
+
+void xbot::service::Scheduler::ThreadLoop() {
+  while (true) {
+    {
+      Lock lk(&state_mutex_);
+      if (stopped_) return;
+    }
+
+    uint32_t now_micros = system::getTimeMicros();
+
+    // TODO: Handle overflow
+
+    if (schedules_head_->next_execution_ > now_micros) {
+      uint32_t sleep_time = schedules_head_->next_execution_ - now_micros;
+      // TODO: Sleep in an interruptible way (new schedule, stopped_ set).
```

Comment:
[chEvtWaitOneTimeout](https://chibios.org/doc/trunk/full_rm/group__events.html#ga6b40fab346fb36d5e1e811d702804773) is probably a good candidate. We can define an event flag for 'schedule changed' and just wait for that